assembly/src/library/mod.rs

```diff
@@ -94,15 +95,79 @@ impl CompiledLibrary {
     }
 }
 
+// KERNEL LIBRARY
+// ================================================================================================
+
+/// Represents a library containing a Miden VM kernel.
+///
+/// This differs from the regular [CompiledLibrary] as follows:
+/// - All exported procedures must be exported directly from the kernel namespace (i.e., `#sys`).
+/// - The number of exported procedures cannot exceed [Kernel::MAX_NUM_PROCEDURES] (i.e., 256).
+pub struct KernelLibrary {
+    kernel: Kernel,
+    kernel_info: ModuleInfo,
+    library: CompiledLibrary,
```

Comment:
I was thinking a `KernelLibrary` would only wrap a `CompiledLibrary` (and would make all its checks in the constructor). Then the `ModuleInfo` and `Kernel` can be derived from the `CompiledLibrary` in methods. The advantage of that approach would be that the struct would be simpler to understand, since it wouldn't store any duplicate information.